common/wforce-geoip2.cc

```diff
@@ -0,0 +1,143 @@
+/*
+ * This file is part of PowerDNS or weakforced.
+ * Copyright -- PowerDNS.COM B.V. and its contributors
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of version 3 of the GNU General Public License as
+ * published by the Free Software Foundation.
+ *
+ * In addition, for the avoidance of any doubt, permission is granted to
+ * link this program with OpenSSL and to (re)distribute the binaries
+ * produced as the result of such linking.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "config.h"
+#ifdef HAVE_MMDB
+#include "wforce-geoip2.hh"
+#include "wforce_exception.hh"
+#include "dolog.hh"
+
+std::mutex geoip2_mutx;
+std::map<std::string, std::shared_ptr<WFGeoIP2DB>> geoip2Map;
+
+WFGeoIP2DB::WFGeoIP2DB(const std::string& filename)
+{
+  int res;
+  memset(&d_db, 0, sizeof(d_db));
+  if ((res = MMDB_open(filename.c_str(), MMDB_MODE_MMAP, &d_db)) < 0)
+    throw WforceException(std::string("Cannot open ") + filename + std::string(": ") + std::string(MMDB_strerror(res)));
+  d_init = true;
+  debuglog("Opened MMDB database %s (type: %s version: %d.%d)", filename, d_db.metadata.database_type, d_db.metadata.binary_format_major_version, d_db.metadata.binary_format_minor_version);
+}
+
+WFGeoIP2DB::WFGeoIP2DB()
+{
+  memset(&d_db, 0, sizeof(d_db));
+}
+
+WFGeoIP2DB::~WFGeoIP2DB()
+{
+  MMDB_close(&d_db);
+}
+
+std::string WFGeoIP2DB::lookupCountry(const ComboAddress& address)
+{
+  MMDB_entry_data_s data;
+  MMDB_lookup_result_s res;
+
+  if (!mmdbLookup(address.toString(), res) ||
+      (MMDB_get_value(&res.entry, &data, "country", "iso_code", NULL) != MMDB_SUCCESS) ||
+      !data.has_data)
+    return {};
+
+  return std::string(data.utf8_string, data.data_size);
+}
+
+std::string WFGeoIP2DB::lookupISP(const ComboAddress& address)
+{
+  MMDB_entry_data_s data;
+  MMDB_lookup_result_s res;
+
+  if (!mmdbLookup(address.toString(), res) ||
+      (MMDB_get_value(&res.entry, &data, "isp", NULL) != MMDB_SUCCESS) ||
+      !data.has_data)
+    return {};
+
+  return std::string(data.utf8_string, data.data_size);
+}
+
+WFGeoIPRecord WFGeoIP2DB::lookupCity(const ComboAddress& address)
```

Comment:
lookupCity? This does not lookup city? =)

Reply:
Well it does lookup city, along with a bunch of other fields, all of which (in combination) are only available in the City DB.

Reply:
Also, I wanted to keep the interface identical to the existing legacy functions (as much as possible)

Reply:
i.e. the XXX in lookupXXX refers more to the DB that is being looked up in (because that's how the legacy API was written)

Reply:
just looks confusing

Reply:
Well that's the way it is, and I'm really loath to change the API cos there's already customers using it, and I want to make their switch to Geoip2 as easy as possible